web/pandas/pdeps/0010-required-pyarrow-dependency.md

```diff
@@ -0,0 +1,151 @@
+# PDEP-10: PyArrow as a required dependency for default string inference implementation
+
+- Created: 17 April 2023
+- Status: Under discussion
+- Discussion: [#52711](https://github.com/pandas-dev/pandas/pull/52711)
+              [#52509](https://github.com/pandas-dev/pandas/issues/52509)
+- Author: [Matthew Roeschke](https://github.com/mroeschke)
+          [Patrick Hoefler](https://github.com/phofl)
+- Revision: 1
+
+## Abstract
+
+This PDEP proposes that:
+
+- PyArrow becomes a runtime dependency starting with pandas 3.0
+- The minimum version of PyArrow supported starting with pandas 3.0 is version 7 of PyArrow.
+- When the minimum version of PyArrow is bumped, PyArrow will be bumped to the highest version that has
+  been released for at least 2 years.
+- Starting in pandas 2.1, pandas raises a ``FutureWarning`` when needing to infer string data that the future
+  data type result will be `ArrowDtype` with `pyarrow.string` instead of object
```

Comment:
> don't we want to go further, and raise a `FutureWarning` upon `import pandas` if `pyarrow` isn't installed, warning that in the future it will become a required dependency?

I agree with Marco here.  I'd also suggest that if we go that route, the message points to a Github issue where we can gather feedback

Reply:
Yeah a feedback issue is a very good idea.

Reply:
I think feedback is also a great idea, but isn't raising a warning on import so soon after just releasing 2.0 for the next major release counterproductive for the whole user experience? Not aware of any other solution but I think this might cause a lot of frustrations.

Reply:
I'm -1 on the warning on import pandas. I assume the majority of pandas users don't care if pyarrow is installed or not, and we'll be annoying them for no good reason.

For pip, I think we can start packaging arrow with the default pandas, and maybe have a pandas[noarrow] for users who want to avoid pyarrow until it's required. Not sure if for conda-forge it's so easy. If we do that, then the warning would make sense to me.

Reply:
We could hold off on the import warning til 2.2 and only raise a warning whenever we would encounter a change in behavior (specifically the string inference). But I guess this would happen to everybody, so not really less noisy

Reply:
Dask did something similar for Arrow, we pull it by default when dask is installed

Reply:
> I'm -1 on the warning on import pandas. I assume the majority of pandas users don't care if pyarrow is installed or not, and we'll be annoying them for no good reason.

My proposal is to only warn users if they import pandas AND they don't have pyarrow installed.  Then the question is how they silence the warning:

1.  (Preferred) - Install pyarrow
2.  Maybe we allow them to set an environment variable to silence it

Ideally, they do (1) and if there are issues, we get a report about it. 

The problem we have is that we don't know how many people will really be affected by requiring pyarrow to be installed.  If it is a small number, then we are creating pain for a large segment of the community who have to silence the warning. But if it is anything over 10% (IMHO), we need to hear from those people, and a warning up top is the best way to get it.

Reply:
Is the status of pyarrow in pandas still experimental?
(Sorry, haven't kept up with this)

For me, warning that we are going to require pyarrow signals that you should be able to use the pyarrow-backed arrays everywhere, and there seems to be quite a bit of buggy/notimplemented things currently.
(at least going by https://github.com/pandas-dev/pandas/issues?q=is%3Aissue+is%3Aopen+label%3ABug+label%3AArrow)

Reply:
Personally, I think I am in favor of starting to warn in 2.2.

Reply:
> My proposal is to only warn users if they import pandas AND they don't have pyarrow installed. 

That's clear, still -1. I think a huge amount of pandas users just install pandas with pip install pandas or same with conda. They won't get pyarrow until they explicitly try to use pyarrow strings to a pyarrow reader, which only advanced users will do. So, almost every regular pandas user will get an intimidating warning just by importing the module, feeling they did something wrong when they didn't. A very poor user experience in my opinion. And I don't see the advantage. With the exception of few users I assume most people will just be happy with whatever pip install pandas requires, and if we internally require numpy, pyarrow or something else, they don't care.

Reply:
> So, almost every regular pandas user will get an intimidating warning just by importing the module, feeling they did something wrong when they didn't.

That depends on how we word the warning.  If we say something like "You better install pyarrow now or everything will break", that will scare them.  If we say something like "Starting with pandas 3.0, pyarrow will become a required installed dependency for pandas. Install it now to identify any potential issues and to remove this warning. Report issues to https://github.com/pandas-dev/pandas/issues/xxxxx"  I don't think the latter is intimidating.

Having said that, I think the specifics of when this warning will appear should be detailed as part of this PDEP.

Reply:
To be fair, "just" warning when inferring string data will result in practically every user seeing the warning anyway, so maybe that's enough (if I understand correctly?)

So, if I install pandas 2.1.0 and don't have pyarrow installed, then `pd.Series(['foo'])` would raise a `FutureWarning` telling me that in the future the default will be a pyarrow string dtype, and that to opt-in to the new behaviour I need to install pyarrow and set `dtype='string[pyarrow]'`? Whereas if I did have pyarrow installed, then the warning would just say to set `dtype='string[pyarrow]'`?

Setting `dtype=` everywhere to silence the warning could be quite a lot of work, maybe there's a simpler way for users to opt-in to this?

Reply:
@MarcoGorelli this is a good point and compromise. Since one of the arguments in favour of this PDEP regards string treatment this seems like a good place to put the warning and relay the message about feedback in a GH issue.
It also allows a pandas global variable to suppress warnings such as this rather than have to rely on environment variable to suppress an 'on import' warning.

Reply:
Right now I am leaning towards @MarcoGorelli's suggestion of only warning when pandas needs to perform string inference in 2.1.  I also think it's a good medium of warning that pyarrow will be required _and_ where it will make a difference in 3.0.